updater/src/updater.rs

```diff
@@ -68,34 +78,69 @@ impl Default for UpdatePolicy {
 			filter: UpdateFilter::None,
 			track: ReleaseTrack::Unknown,
 			path: Default::default(),
+			max_delay: 100,
+			frequency: 20,
 		}
 	}
 }
 
+/// The current updater status
+#[derive(Clone, Debug, PartialEq)]
+enum UpdaterStatus {
+	/// Updater is currently disabled.
+	Disabled,
+	/// Updater is currently idle.
+	Idle,
+	/// Updater is waiting for block number to fetch a new release.
+	Waiting {
+		release: ReleaseInfo,
+		binary: H256,
+		block_number: BlockNumber,
+	},
+	/// Updater is fetching a new release.
+	Fetching {
+		release: ReleaseInfo,
+		binary: H256,
+		retries: u32,
+	},
+	/// Updater failed fetching a new release and it is now backing off until the next retry.
+	FetchBackoff {
+		release: ReleaseInfo,
+		binary: H256,
+		backoff: (u32, Instant),
+	},
+	/// Updater is ready to update to a new release.
+	Ready {
+		release: ReleaseInfo,
+	},
+	/// Updater has installed a new release and can be manually restarted.
+	Installed {
+		release: ReleaseInfo,
+	},
+}
+
+impl Default for UpdaterStatus {
+	fn default() -> Self {
+		UpdaterStatus::Idle
+	}
+}
+
 #[derive(Debug, Default)]
 struct UpdaterState {
 	latest: Option<OperationsInfo>,
-
-	fetching: Option<ReleaseInfo>,
-	ready: Option<ReleaseInfo>,
-	installed: Option<ReleaseInfo>,
-
 	capability: CapState,
-
-	disabled: bool,
-
-	backoff: Option<(u32, Instant)>,
+	status: UpdaterStatus,
 }
 
 /// Service for checking for updates and determining whether we can achieve consensus.
-pub struct Updater {
+pub struct Updater<O = OperationsContractClient, F = fetch::Client> {
 	// Useful environmental stuff.
 	update_policy: UpdatePolicy,
-	weak_self: Mutex<Weak<Updater>>,
+	weak_self: Mutex<Weak<Updater<O, F>>>,
```

Comment:
can we somehow get rid of this? I know that this will require moving some logic to a separate structure, but `weak_self ` even sounds terribly :p